### ISSUE TITLE
EF Core 3.1.1 - FromSQLRaw - multiple column names with same name in select from different tables in sql statement

### ISSUE THREAD

**Issue by sacerande** [dotnet/efcore#21592]
I am migrating application from .Net framework to .Net core. there I have some .net entity framework code, which uses raw sql queries and maps it to entity models.
similarly I am trying to achieve with ef core.

In my code, following sql  query gets is generated
```

    SELECT gt.*,RefTable1.*,RefTable2.* 
    FROM APP_USER_TO_PGM_PERMISSION AS gt 
    INNER JOIN APP_USERS  AS  RefTable1 ON gt.USER_ID=RefTable1.ID  
    INNER JOIN APP_LK_MODULEPERMISSIONS  AS  RefTable2 
    ON gt.PERMISSION_ID=RefTable2.ID
```

after call to following statement, 
`    var testLst = db.Set<APP_USER_TO_PGM_PERMISSION>().FromSqlRaw(strQuery).AsNoTracking().ToList(); `

`Exception "$exception	{"An item with the same key has already been added. Key: ID"}	System.ArgumentException" is thrown.`

```
    stack trace:
     at System.ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException[T](T key)
       at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
       at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.BuildIndexMap(IReadOnlyList`1 columnNames, DbDataReader dataReader)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.InitializeReader(DbContext _, Boolean result)
       at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
```

**_Observed - EF Core mapper doesn't support multiple columns with same name even if they belong to different tables. It was working with Entity framework._**

Method in EF Core code where it fails

```
    public static int[] BuildIndexMap([CanBeNull] IReadOnlyList<string> columnNames, [NotNull] DbDataReader dataReader)
    {
        var readerColumns = Enumerable.Range(0, dataReader.FieldCount)
            .ToDictionary(dataReader.GetName, i => i, StringComparer.OrdinalIgnoreCase);

        var indexMap = new int[columnNames.Count];
        for (var i = 0; i < columnNames.Count; i++)
        {
            var columnName = columnNames[i];
            if (!readerColumns.TryGetValue(columnName, out var ordinal))
            {
                throw new InvalidOperationException(RelationalStrings.FromSqlMissingColumn(columnName));
            }

            indexMap[i] = ordinal;
        }

        return indexMap;
    }
```

### Steps to reproduce
just create a sql query as above with multiple columns with same name but from different tables and try to execute FromSqlRaw as shown in details..

multiple column names with same name from different table are not supported. which was supported in Entity framework.

### Further technical details

EF Core version:3.1.1
Database provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer):Microsoft.EntityFrameworkCore.SqlServer
Target framework: (e.g. .NET Core 3.1)
Operating system:
IDE: (e.g. Visual Studio 2019 16.3)

**Comment by sacerande** [comment]
![image](https://user-images.githubusercontent.com/41945701/87221259-31ff3980-c388-11ea-870e-4615e41a144d.png)

**Comment by smitpatel** [comment]
And how do you expect EF Core to figure out which column to read value from for those duplicated column names?

**Comment by sacerande** [comment]
It could be same as entity framework. Table alias is there in select query to understand which table column belongs to.
If it's not possible at least the first one should be preferred and matched, without throwing exception.
I am not sure ef core still has a support for mapping sub collection properties.

**Comment by smitpatel** [comment]
> Table alias is there in select query to understand which table column belongs to.

That requires parsing SQL and even then, you are asking us to materialize `APP_USER_TO_PGM_PERMISSION` entity, how would we know which "table" is actually right one. 
You can easily wrap your SQL in a subquery to only Select correct column.
Since there are 2 potential matches, I think it is better to throw exception and ask user to be explicit about it rather than selecting first because if it is not the correct one then it will corrupt your data.

**Comment by sacerande** [comment]
Yes right (SQL parsing) feature must be powerful like entity framework. 
as we specify APP_USER_TO_PGM_PERMISSION  it in generic type T of method, it indicates that it is top level entity which we are trying to generate.

It is not good practice to write 50+ columns in query explicitly just because ef core rawsqlquery is not powerful enough to generate entity/or any other type we specify.
We have to change product at too many places because of lack of this feature. EF was allowing us to do this (SQL, MySQL, Oracle) for the 3 dbs it worked, but Ef core lacks it.

**Comment by sacerande** [comment]
To which type you should match is already known.
Type's name we specify or if type is entity then it is table name of entity type.
Set<APP_USER_TO_PGM_PERMISSION>

**Comment by sacerande** [comment]
There are 2 same matches becsuse you are not reading the content by alias. and that is flaw we see duplicates but are not duplicate.
So gt.ID, refTable1.ID are 2 different things.
You see it as one because you consider only "ID" but not alias which every major relational db supports.

**Comment by smitpatel** [comment]
`ID` is the alias and we read by using alias only.

![image](https://user-images.githubusercontent.com/1528107/87243221-24c66700-c3e9-11ea-9de9-e69493d28c7f.png)

**Comment by sacerande** [comment]
Whatever...then why your dictionary is of unique column name ? I am interested in results..if EF core can't give simple feature like what we have in entity framework this is really strange...just tell me yes/no you can accept and implement the feature in EF core?
Repeating again Entity framework was allowing me this operation using raw SQL queries.

If you say NO I am going to remove ef core package and going with pure ado.net and my mapper classes. I am going to share the same thread with everyone, as I am really not getting basic thing implemented in EF Core. Strange!

**Comment by sacerande** [comment]
@smitpatel  or someone else what's your decision based on information and comments made by me..just tell me Yes/No...

**Comment by smitpatel** [comment]
Are you asking us to parse SQL?

**Comment by smitpatel** [comment]
> Whatever...then why your dictionary is of unique column name ?

While column name does not have to be unique but the name of the properties in the entity we are materializing has to be unique. So I am not sure why what SQL you can write affect, what we _can_ materialize.

**Comment by sacerande** [comment]
@smitpatel I have given more than sufficient information to you. please do run the same query with entity framework, it does give me result without exception and "",""

**Comment by sacerande** [comment]
I lost my hope now, removing ef core package and not any further bug logging here, and sharing the same across.

**Comment by ErikEJ** [comment]
Could you share an example of how this worked with EF6?

**Comment by sacerande** [comment]
Hi Erik,
please refer following .net framework sample app that uses EF 6, and fetches some data from sql server. 
https://github.com/sacerande/StudentAppDotNetFramework
please refer readme.md file it contains some instructions on how to setup local database.

once you are done,
please refer following, it contains same code, but with ef core. when run it fails with the error mentioned in this thread.
https://github.com/sacerande/StudentAppWithDotNetCore

Thanks.

**Comment by ajcvickers** [comment]
I investigated this by running the EF6 code from the repro:

```C#
var lst = db.student.SqlQuery("select tab1.*, tab2.* from student tab1 inner join subjects tab2 on tab1.id = tab2.studid").ToList();
foreach(var stud in lst)
{
    Console.WriteLine("----------------------------");
    Console.WriteLine($"Student FirstName {stud.firstname} LastName {stud.lastname} has following subjects...");
    foreach(var sub in stud.subjects)
    {
        Console.WriteLine($"Subject ID: {sub.id}, Subject Name: {sub.subName}");
    }
    Console.WriteLine("----------------------------");
}
```
The reference to `stud.subjects` is causing EF to lazy-load each time through the loop. This produces the following output:

```
----------------------------
Student FirstName S LastName E has following subjects...
Subject ID: 2000, Subject Name: H
Subject ID: 2001, Subject Name: M
----------------------------
----------------------------
Student FirstName S LastName E has following subjects...
Subject ID: 2000, Subject Name: H
Subject ID: 2001, Subject Name: M
----------------------------
----------------------------
Student FirstName S1 LastName E1 has following subjects...
Subject ID: 2002, Subject Name: G
----------------------------
```

Turning off lazy-loading results in no `subjects` being loaded:

```
----------------------------
Student FirstName S LastName E has following subjects...
----------------------------
----------------------------
Student FirstName S LastName E has following subjects...
----------------------------
----------------------------
Student FirstName S1 LastName E1 has following subjects...
----------------------------
```

**Comment by smitpatel** [comment]
What are the duplicate columns in FromSql?

**Comment by ajcvickers** [comment]
@smitpatel The duplicates are being resolved into a single instance but are still returned multiple times in the result:
![image](https://user-images.githubusercontent.com/1430078/88432916-fdd93f00-cdb1-11ea-9ce6-3f01764de193.png)

**Comment by smitpatel** [comment]
That is not the issue. We do that already based on tracking settings.
 The issue is the SQL contains multiple columns with same name. And we have to figure out for each property in the projection, the index of corresponding column which we do based on name.
if the query is 
```
customers.FromSql("Select c.City, o.City, l.City from Customers as C join Orders as o on c.Id = o.Id join Lands as l on c.Id = l.Id)
    .Select(c => new { c.City})
```
Which city are we reading?

**Comment by ajcvickers** [comment]
@smitpatel Can't tell that just from the logging. It would require debugging into the EF6 query pipeline.

**Comment by smitpatel** [comment]
A fake SQL could also work. as long as it is having multiple same named columns.
````
Select c.City, o.City
From (
    Select N'Seattle' as City
) Join (
    Select N'New York' as City
) on 1 = 1
````

**Comment by ajcvickers** [comment]
@smitpatel SQL rejected that SQL with syntax error near `on`. However, given the original query:

```sql
select tab1.*, tab2.* from student tab1 inner join subjects tab2 on tab1.id = tab2.studid
```

If I change it to:

```sql
select tab1.firstname, tab1.lastname, tab1.id, tab2.id, tab2.studid, tab2.subName from student tab1 inner join subjects tab2 on tab1.id = tab2.studid
```
then `student.id` gets populated from `tab1.id`.

However, if I switch the order of `tab1.id` and `tab2.id`:

```sql
select tab1.firstname, tab1.lastname, tab2.id, tab1.id, tab2.studid, tab2.subName from student tab1 inner join subjects tab2 on tab1.id = tab2.studid
```

then `student.id` gets populated from `tab2.id`.

So it looks like EF6 is taking the first column with a name that matches.

**Comment by smitpatel** [comment]
Let's discuss in team meeting if that is what something we should do or we should continue throwing error. It is very easy to get incorrect data without even realizing.

**Comment by ajcvickers** [comment]
Discussed again and agreed that we should not change the behavior here.

**Comment by sacerande** [comment]
Don't change it, but can you enhance it?
Eg. with default parameter to method like StrictMatch = true, by default if true then throw exception. If false try to match first one, similar to how EF does.
If there are 50+ columns in my table, I don't think it's good idea to write them in query, just because you do not support what EF6 was doing.
Design can be enhanced to support, because it is really a question if you do not support 
By design EF6 is wrong OR EFCore is wrong ?
Giving first match vs throwing exception, which one is good for companies migrating from EF to EFCore?
Is it a good idea to ask everyone to write all columns from different tables at too many places because you do not support it, is it customer friendly?

**Comment by ajcvickers** [comment]
@sacerande This is either a bug or a poorly designed behavior in EF6. Either way, it's not something that is going to change in EF6, but it's also not a behavior that it makes sense to bring to EF Core.

**Comment by v-zubritsky** [comment]
Hi folks. I have just encountered this issue as well, while migrating from EF 6 to EF Core 3.1.12 and I believe that at least error message could be improved here. 
Generic "An item with the same key has already been added" is not informative and makes it look like it's a bug in EF and not the expected case. Also it gives no information about what column exactly was duplicated, while having it could be nice.
Thanks.
